### PR TITLE
Feature/mapbook exports

### PR DESCRIPTION
**Pull request by andrewphilipsmith**
Not ready to merge yet.

Need to test all of the changes to references to .Net 4.

### CODE REVIEW

**Comment by andrewphilipsmith** [comment]
@kes1 I've just been looking at this. When I try and compile it I get an error on this file; `frmExportMain.cs` on line 26:

```using mapbook_export_tools;```

`frmExportMain.cs(26,7): error CS0246: The type or namespace name 'mapbook_export_tools' could not be found (are you missing a using directive or an assembly reference?) [d:\jenkins\workspace\MapExportTool\arcgis10_mapping_tools\MapActionToolbars\MapActionToolbars.csproj]`

There are no other occurrences of the phrase `mapbook_export_tools` in the solution. This makes me suspect that there are some files missing from the repo. Please could you check that all the files (particularly any new files you've created have been added to this branch? Let me know if you need any help with this. Thanks Andy.

**Comment by kes1** [comment]
Hi @andrewphilipsmith - makes sense, `mapbook_export_tools` is the class that's generated from the geoporcessing toolbox.  

I probably need to add it to the repo, will double check later.

File: arcgis10_mapping_tools/MapActionToolbars/mapbook_export_tools.dll.config [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0"?>
+<configuration>
+  <toolbox directory="C:\Users\andrew.KESTERTON\Documents\GitHub\mapaction-toolbox\mapbook_export_tool" />
```
Comment (andrewphilipsmith) [inline arcgis10_mapping_tools/MapActionToolbars/mapbook_export_tools.dll.config]: @kes1 I think this could be the culprit. That path is unique to you.

**Comment by andrewphilipsmith** [comment]
Hi @kes1. I had a look through this last night. ESRI have setup a real maze to navigate here!

I used the "ArcGIS Toolbox Reference..." wizard in Visual Studio and that fix the reference an allow me to build the solution locally. Although this works locally, it kinda makes me nervous because it looks like you can't [build your application in a single step](https://www.joelonsoftware.com/2000/08/09/the-joel-test-12-steps-to-better-code/).

However with a bit of digging around it seems that you can reproduce the actions of this wizard using this commandline tool:
`C:\Program Files (x86)\ArcGIS\DeveloperKit10.2\IDEIntegration\VisualStudio\bin\GPAssemblyBuilder.exe`

I think it would be possible to add a Re-build Event Commandline into the MapAcitonToolbars Project, which would regenerate the .dll on each compile. 

That gets at least partly out of the woods. However we still have the binary .tbx file in the source tree. Under what circumstances would the toolbox (.tbx) file need to be updated?
* If the underlying python was changed?
* If just the list of parameters were updated?

Basically is it possible for the .tbx file to get out of sync with the python files? If so then we still have the risk of odd behaviours creeping. I'm not sure how significant this risk would be or how we would manage it.

I hope that's helpful. They don't make this easy do they?

**Comment by kes1** [comment]
Hi @andrewphilipsmith that's a great find! I was about to start digging now.. I'll try and get a re-build event setup. Does the build server have the ArcGIS SDK installed, and at the same path?

Re the toolbox, I think it has to be refreshed whenever the script changes. I initially tried with a python toolbox which is a much better solution, but these can't be built into a Geoprocessing Assembly, suspect that may change with time though.

**Comment by andrewphilipsmith** [comment]
Unfortunately the `GPAssemblyBuilder.exe` tool didn't work as I had hoped. I don't know why, but then there isn't any recommendation from ESRI that we should use this anyway so I guess it was a bit of a long shot.

I merged this branch `feature/mapbook-exports` with another I had been working on `map-template-tests` (PR#143) in order to pick up on a couple of relevant tests I had been working on. I'll add more comments there.
